### PR TITLE
Fix OutputNode, Catalyst, Recipe bugs, Add shopping list

### PR DESCRIPTION
**Pull request by tvwenger**
Fixes the Glass Panel M recipe error in https://github.com/lgfrbcsgo/du-factory-generator/issues/50 and the bugs reported in https://github.com/lgfrbcsgo/du-factory-generator/issues/51 and https://github.com/lgfrbcsgo/du-factory-generator/issues/52. Adds the shopping list feature in https://github.com/lgfrbcsgo/du-factory-generator/issues/49

### CODE REVIEW

**Comment by tvwenger** [comment]
Implements https://github.com/lgfrbcsgo/du-factory-generator/issues/54 by showing production rate when user selects number of assemblers. Also fixes a bug with `startingFactory` - it was being overwritten after adding to an existing factory.

File: src/factory.ts [hunk 1]
```diff
@@ -59,33 +59,33 @@ function produce(item: Item, rate: PerMinute, factory: FactoryGraph): ContainerN
...
+        // Check that we can add incoming and outgoing link to this container,
+        // and that we can add outgoing links to the catalyst transfer unit output
+        // if any
         if (
             container.canAddIncomingLinks(additionalIndustries) &&
             container.canAddOutgoingLinks(1) &&
-            catalystCheck
+            (catalystTransferUnit === undefined ||
+                (catalystTransferUnit.output !== undefined &&
+                    catalystTransferUnit.output.canAddOutgoingLinks(additionalIndustries)))
```
Comment (lgfrbcsgo) [inline src/factory.ts:83]:
```suggestion
            (catalystTransferUnit === undefined ||
                catalystTransferUnit.output?.canAddOutgoingLinks(additionalIndustries) === true)
```
You can use the optional chaining feature to get rid of that additional check. Basically, if `catalystTransferUnit.output` is `undefined`, the entire expression will return `undefined`. `catalystTransferUnit.output?.canAddOutgoingLinks(additionalIndustries)` has return type of `boolean | undefined`.

File: src/factory.ts [hunk 2]
```diff
@@ -112,6 +112,10 @@ function produce(item: Item, rate: PerMinute, factory: FactoryGraph): ContainerN
     for (let i = 0; i < additionalIndustries; i++) {
         const industry = factory.createIndustry(item)
         industry.outputTo(outputs[i % outputs.length])
+        // link catalyst byproduct
+        if (catalystTransferUnit !== undefined && catalystTransferUnit.output !== undefined) {
```
Comment (lgfrbcsgo) [inline src/factory.ts:116]:
```suggestion
        if (catalystTransferUnit?.output !== undefined) {
```
🙂

File: src/factory.ts [hunk 3]
```diff
@@ -509,24 +541,56 @@ export function buildFactory(
...
+                if (isTransferNode(consumer) && isCatalyst(consumer.item)) {
+                    catalystTransferUnit = consumer
+                }
+            }
+
+            if (
+                checkOutput.canAddIncomingLinks(count) &&
+                (catalystTransferUnit === undefined ||
+                    (catalystTransferUnit.output !== undefined &&
+                        catalystTransferUnit.output.canAddOutgoingLinks(count)))
```
Comment (lgfrbcsgo) [inline src/factory.ts:565]:
```suggestion
                    catalystTransferUnit?.output.canAddOutgoingLinks(count) === true)
```
🙂

File: src/factory.ts [hunk 4]
```diff
@@ -535,19 +599,45 @@ export function buildFactory(
...
             const industry = factory.createIndustry(item)
-            industry.outputTo(output)
+            for (const output of outputs) {
+                if (output.canAddIncomingLinks(1)) {
+                    industry.outputTo(output)
+                    break
+                }
+            }
+            // link transfer unit output to industry, if any
+            if (catalystTransferUnit !== undefined && catalystTransferUnit.output !== undefined) {
```
Comment (lgfrbcsgo) [inline src/factory.ts:638]:
```suggestion
            if (catalystTransferUnit?.output !== undefined) {
```
🙂

**Review by lgfrbcsgo — APPROVED** [review]
Approved.

Could you please create separate PRs for each issue next time? It is really hard to figure out which changes belong to which issue.

**Pull request merged**